plutus-tx-plugin/test/Recursion/Spec.hs

```diff
@@ -27,38 +28,49 @@ tests =
   testNested "Recursion" . pure $
     testNestedGhc
       [ goldenUPlcReadable "length-direct" compiledLengthDirect
-      , goldenPirReadable "length-direct" compiledLengthDirect
       , goldenEvalCekCatch
           "length-direct"
-          [compiledLengthDirect `unsafeApplyCode` liftCodeDef [1, 2, 3]]
+          [compiledLengthDirect `unsafeApplyCode` liftCodeDef [1..10]]
       , goldenBudget
           "length-direct"
-          (compiledLengthDirect `unsafeApplyCode` liftCodeDef [1, 2, 3])
+          (compiledLengthDirect `unsafeApplyCode` liftCodeDef [1..10])
       , goldenUPlcReadable "length-fix" compiledLengthFix
-      , goldenPirReadable "length-fix" compiledLengthFix
       , goldenEvalCekCatch
           "length-fix"
-          [compiledLengthFix `unsafeApplyCode` liftCodeDef [1, 2, 3]]
+          [compiledLengthFix `unsafeApplyCode` liftCodeDef [1..10]]
       , goldenBudget
           "length-fix"
-          (compiledLengthFix `unsafeApplyCode` liftCodeDef [1, 2, 3])
+          (compiledLengthFix `unsafeApplyCode` liftCodeDef [1..10])
+      , goldenUPlcReadable "length-peeled" compiledLengthPeeled
+      , goldenEvalCekCatch
+          "length-peeled"
+          [compiledLengthPeeled `unsafeApplyCode` liftCodeDef [1..10]]
+      , goldenBudget
+          "length-peeled"
+          (compiledLengthPeeled `unsafeApplyCode` liftCodeDef [1..10])
       ]
 
 lengthDirect :: [Integer] -> Integer
-lengthDirect xs = case xs of
+lengthDirect = \case
   []       -> 0
-  (_ : ys) -> 1 PlutusTx.+ lengthDirect ys
+  (_ : xs) -> 1 PlutusTx.+ lengthDirect xs
 
 lengthFix :: [Integer] -> Integer
 lengthFix =
   fix
-    ( \f xs -> case xs of
+    ( \f -> \case
         []       -> 0
-        (_ : ys) -> 1 PlutusTx.+ f ys
+        (_ : xs) -> 1 PlutusTx.+ f xs
     )
 
+lengthPeeled :: [Integer] -> Integer
+lengthPeeled = $$(peel 3 (\f -> [|| \case [] -> 0; (_ : xs) -> 1 PlutusTx.+ $$f xs ||]))
```

Comment:
Unfortunately this implementation of `lengthPeeled`  needs to re-implement the `lengthFix` explicitly repeating its step-function code inside the TH quotes.

While this is already a useful macro and for some developers/cases the code repetition is not a big deal, 
I think its possible to improve by implementing it in the plugin with its access to the unfolding of `lengthFix` by
extracting the step-function automatically in order to re-compose it the way `peel` does.